qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -404,3 +407,73 @@ def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **
+
+    This plugin name is :``permutation.token_swapper`` which can be used as the key on
+    an :class:`~.HLSConfig` object to use this method with :class:`~.HighLevelSynthesis`.
+
+    In more detail, this plugin is used to synthesize objects of type `PermutationGate`.
+    When synthesis succeeds, the plugin outputs a quantum circuit consisting only of swap
+    gates. When synthesis does not succeed, the plugin outputs `None`.
+
+    If either `coupling_map` or `qubits` is None, then the synthesized circuit
+    is not required to adhere to connectivity constraints, as is the case
+    when the synthesis is done before layout/routing.
+
+    On the other hand, if both `coupling_map` and `qubits` are specified, the synthesized
+    circuit is supposed to adhere to connectivity constraint. At the moment, the plugin
+    only works when `qubits` represents a connected subset of `coupling_map` (if this is
+    not the case, the plugin outputs `None`).
+
+    The plugin supports the following plugin-specific options:
+
+    * trials: The number of trials for the token swapper to perform the mapping. The
+      circuit with the smallest number of SWAPs is returned.
+    * seed: The argument to the token swapper specifying the seed for random trials.
+    * parallel_threshold: The argument to the token swapper specifying the number of nodes
+      in the graph beyond which the algorithm will use parallel processing.
+
+    For more details on the token swapper algorithm, see to the paper:
+    `arXiv:1809.03452 <https://arxiv.org/abs/1809.03452>`__.
+
+    """
+
+    def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **options):
+        """Run synthesis for the given Permutation."""
+
+        trials = options.get("trials", 5)
+        seed = options.get("seed", 0)
+        parallel_threshold = options.get("parallel_threshold", 50)
+
+        pattern = high_level_object.pattern
+        pattern_as_dict = {j: i for i, j in enumerate(pattern)}
+
+        # When the plugin is called from the HighLevelSynthesis transpiler pass,
+        # the coupling map already takes target into account.
+        if coupling_map is None or qubits is None:
+            # The abstract synthesis uses a fully connected coupling map, allowing
+            # arbitrary connections between qubits.
+            used_coupling_map = CouplingMap.from_full(len(pattern))
+        else:
+            # The concrete synthesis uses the coupling map restricted to the set of
+            # qubits over which the permutation gate is defined. Currently, we require
+            # this set to be connected (otherwise, replacing the node in the DAGCircuit
+            # that defines this PermutationGate by the DAG corresponding to the constructed
+            # decomposition becomes problematic); note that the method `reduce` raises an
+            # error if the reduced coupling map is not connected.
```

Comment:
We should be able to handle disconnected subgraphs if we work with the graph object directly. Something like:

```python
graph = coupling_map.graph.subgraph(qubits).to_undirected()
```

The only issue with that is I don't think there is any guarantees on the index ordering from `PyDiGraph.subgraph()`. So you might need to do:

```python
for i in coupling_map.graph.node_indices():
    coupling_map.graph[i] = i
graph = coupling_map.graph.subgraph(qubits).to_undirected()
```

then you you can use `graph[j]` to get the original index `i`. That being said I don't know what the approximate token swapper does if it can't fulfill the permutation because the graph is disconnected, so this might be the correct behavior anyway.

Reply:
Thanks, this is a very interesting suggestion. By experimenting with the approximate token swapper, it seems that if there is a way to fulfill the permutation (on a disconnected subgraph), it will do so; but if there is no such way, it will throw a rather strange error ``pyo3_runtime.PanicException: IndexMap: key not found``.

```
graph = rx.PyGraph()
graph.extend_from_edge_list([(0, 1), (2, 3)])
swapper = ApproximateTokenSwapper(graph)
swapper.map({1: 0, 0: 1, 2: 3, 3: 2}, 10)  # this works fine
swapper.map({2: 0, 1: 1, 0: 2, 3: 3}, 10)  # this throws 
```

Question: wrapping the second in the `try`-`except` block, outputs the following message (in pycharm):
```
thread '<unnamed>' panicked at 'IndexMap: key not found', C:\Users\274191756\Desktop\QiskitDevelopment\rustworkx\rustworkx-core\src\token_swapper.rs:211:16
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
Would this be a problem?

Another question: is it safe to modify the original coupling map: `coupling_map.graph[i] = i`, or would it be best to make a copy of that first?

Reply:
Hah, well that's a bug in rustworkx. A `PanicException` is the catch all for a Rust panic or an unhandled error and we shouldn't be panicking with an otherwise valid input. Catching a `PanicException` isn't super straightforward because of design decision in PyO3 it inherits from `BaseException` (which is the parent to `Exception`) to make it on the same level as other low level exceptions that aren't generally recoverable. The thinking being a panic is explicitly an unhandled error so nothing in Python should be able to deal with it. So I wouldn't try to catch it, we'll have to fix it in rustworkx.

As for `coupling_map.graph[i] = i` we don't explicitly reserve the use of the node weights for anything and if something does depend on them (like `CouplingMap.connected_components()`) will overwrite it. Although doing a graph copy is probably fairly lightweight as it's just a rust clone internally. So something like `coupling_map.graph.copy()` is probably a good idea.

Reply:
I pushed up a fix for this in rustworkx: https://github.com/Qiskit/rustworkx/pull/971

Reply:
I really want to support disjoint coupling maps in this and other synthesis plugins.

First, as explained in https://github.com/Qiskit/qiskit/pull/10657#discussion_r1298350869, using `coupling_map.graph.subgraph(qubits)` instead of `coupling_map.reduce(qubits)` does not preserve the index ordering. This leads to a somewhat cumbersome solution that involves remapping the desired permutation to be over the subgraph's qubits (this is the input to the token swapper) and remapping the computed swaps (the output of token swapper) to be over the original qubits. This can be done, but the code is a bit tricky. Furthermore, every other synthesis plugin that is able to support disjoint coupling maps would also need to implement a similar tricky solution.

Do I understand correctly that disconnected coupling maps are now allowed in Qiskit? So an alternative would be to add an argument (something like `allow_disconnected`) to `CouplingMap.reduce` that would not error on disconnected coupling maps. 

There are a few more minor changes requires for `CouplingMap.reduce`. As an example, let's say that the original map is a ring over 8 qubits, and we want to `reduce` to `qubits=(1, 3)`, the important thing is that some of the qubits become edgeless in the reduced map. The current code would not add such edgeless qubits to the reduced coupling map (as it adds new qubits when adding edges), but we should in fact add these. 

@mtreinish, if you agree with the `CouplingMap.reduce` suggestions above, would it make sense to add these as a part of this PR or a separate PR?

The second problem is that there is still one edge case when `TokenSwapper` panics when handling disconnected coupling maps (in addition to https://github.com/Qiskit/rustworkx/pull/971):
```
def test_disjoint_graph(self):
    graph = rx.PyGraph()
    graph.add_node(0)
    graph.add_node(1)
    graph.add_node(2)
    graph.add_node(3)
    swaps = rx.graph_token_swapper(graph, {1: 0, 0: 1, 2: 3, 3: 2}, 10, seed=42)
```
This panics as per
```
thread '<unnamed>' panicked at 'index out of bounds: the len is 0 but the index is 0', C:\Users\274191756\Desktop\QiskitDevelopment\rustworkx\rustworkx-core\src\token_swapper.rs:304:40
```
I would be happy to look into this. One additional question is whether we need to wait till the new version of rustworkx is released (including PR 971 and possibly additional fixes) before this PR could be merged?